### ISSUE TITLE
Capture file from trim-to-single-frame is larger than original capture

### ISSUE THREAD

**Issue by davidlunarg** [LunarG/gfxreconstruct#1164]
I created a capture of Balgdur's Gate 3 on my nVidia Windows system. And I created a trimmed capture of frame 180. Then I replayed both captures:

    $ gfxrecon-replay bgate-new3_20230622T163918.gfxr
    [gfxrecon] WARNING - Incomplete block at end of file
    Total time: 3.258113 seconds
    Replay FPS: 72.741492 fps, 3.258113 seconds, 237 frames, framerange 1-237

    $ gfxrecon-replay trimmed_frame_180_20230622T234755.gfxr
    [gfxrecon] INFO - Loading state for captured frame 180
    [gfxrecon] INFO - Finished loading state for captured frame 180
    Load time:  3.397846 seconds
    Total time: 3.408487 seconds
    Replay FPS: 93.976130 fps, 0.010641 seconds, 1 frame, framerange 180-180

[The warning can be ignored. This is due to the app killing its render thread while a Vulkan API call is in progress.]

The size of the trimmed file is larger than the size of the original:

    $ ls -l bgate-new3_20230622T163918.gfxr  trimmed_frame_180_20230622T234755.gfxr
    -rwxr-xr-x+ 1 David None 434342935 Jun 22 16:41 bgate-new3_20230622T163918.gfxr
    -rwxr-xr-x+ 1 David None 543842095 Jun 22 16:47 trimmed_frame_180_20230622T234755.gfxr

The trimmed file being larger than the original capture doesn't seem right.

**Comment by davidlunarg** [comment]
I looked at a trim-to-single-frame capture to see where the extra size is coming from. [If
you want to skip over the details for now, you can jump to the last paragraph for a summary.]

I started with a capture of Baldur's Gate 3 that had been trimmed to frames 1-175:

    $ ll  trimmed_frames_1_through_175_20230626T213751.gfxr
    -rwxr-xr-x+ 1 David None 397682239 Jun 26 14:38 trimmed_frames_1_through_175_20230626T213751.gfxr

I then trimmed this capture down to a single frame:

    $ VK_LOADER_LAYERS_ENABLE=VK_LAYER_LUNARG_gfxreconstruct \
      GFXRECON_CAPTURE_FILE=trimmed-1-175.gfxr \
      GFXRECON_PAGE_GUARD_ALIGN_BUFFER_SIZES=true \
      GFXRECON_CAPTURE_FRAMES=175 \
      gfxrecon-replay.exe trimmed_frames_1_through_175_20230626T213751.gfxr

Here's the resulting single frame capture file:

    $ ll trimmed-1-175_frame_175_20230626T214533.gfxr
    -rwxr-xr-x+ 1 David None 581940636 Jun 26 14:45 trimmed-1-175_frame_175_20230626T214533.gfxr

As you can see, the capture file containing a single frame is about 180M bytes larger than the capture
file containing 175 frames.

I converted both of the capture files to jsonl files, and also dumped the binaries in those files:

    $ gfxrecon-convert --format jsonl --include-binaries trimmed_frames_1_through_175_20230626T213751.gfxr
    $ gfxrecon-convert --format jsonl --include-binaries trimmed-1-175_frame_175_20230626T214533.gfxr

Examining the dumped binaries shows that that single-frame capture has lots more binary data:

    $ du -sh trimmed_frames_1_through_175_20230626T213751
    1.2G    trimmed_frames_1_through_175_20230626T213751

    $ du -sh trimmed-1-175_frame_175_20230626T214533
    2.5G    trimmed-1-175_frame_175_20230626T214533

The difference in binary sizes is significantly more than the difference in capture file sizes. This is not
surprising - the capture file is compressed, but the dumped binaries are not compressed.

The binary directories show where the differences are.  There are many InitBufferCommand binaries in the capture
with one frame, there are none in the capture with 175 frames:

    $ ls trimmed-1-175_frame_175_20230626T214533/*init_buffer*bin | wc -l :
    1016
    $ du -ch trimmed-1-175_frame_175_20230626T214533/*init_buffer*.bin | grep total :
    553M	total
    $ ls trimmed_frames_1_through_175_20230626T213751/*init_buffer*bin | wc -l :
     ls: cannot access 'trimmed_frames_1_through_175_20230626T213751/*init_buffer*bin': No such file or directory
    $ du -ch trimmed_frames_1_through_175_20230626T213751/*init_buffer*.bin | grep total :
    du: cannot access 'trimmed_frames_1_through_175_20230626T213751/*init_buffer*.bin': No such file or directory

There are many InitImageCommand binaries in the capture with one frame, there are
none in the capture with 175 frames:

    $ ls trimmed-1-175_frame_175_20230626T214533/*init_image*bin | wc -l :
    549
    $ du -ch trimmed-1-175_frame_175_20230626T214533/*init_image*.bin | grep total :
    1.9G	total
    $ ls trimmed_frames_1_through_175_20230626T213751/*init_image*bin | wc -l :
    ls: cannot access 'trimmed_frames_1_through_175_20230626T213751/*init_image*bin': No such file or directory
    $ du -ch trimmed_frames_1_through_175_20230626T213751/*init_image*.bin | grep total :
    du: cannot access 'trimmed_frames_1_through_175_20230626T213751/*init_image*.bin': No such file or directory

However, there are more FillMemoryCommand binaries in the capture with 175 frames than in the capture with one frame:

    $ ls trimmed-1-175_frame_175_20230626T214533/*fill_memory*bin | wc -l :
    61
    $ du -ch trimmed-1-175_frame_175_20230626T214533/*fill_memory*.bin | grep total :
    2.4M	total

    $ ls trimmed_frames_1_through_175_20230626T213751/*fill_memory*bin | wc -l :
    5169
    $ du -ch trimmed_frames_1_through_175_20230626T213751/*fill_memory*.bin | grep total :
    1.2G	total

So the difference is due to the fact that a trim to a single frame results in:
    553M additional bytes for InitBufferCommands
    1.9G additional bytes for InitImageCommands
    1.2G less bytes for FillMemoryCommands
for a net difference of ~1.3Gbytes.

**Comment by panos-lunarg** [comment]
This agrees with my conclusions on the same thing when I investigated the same issue some time ago.
My conclusion was based more on code inspection than on inspecting the capture file.

When capturing un-trimmed and using the `page guard` memory tracking mechanism, GFXR dumps to file only what the user writes to the mapped device memories. In the capture file the these writes are visible as `FillMemoryCommand`s. In a simple use case scenario these are texture and meshes uploads in the init phase and then per frame there are the UBO updates. The content of the rest of the allocated device memories do not to be dumped as it can be regenerated by the gpu on replay.

When capturing trimmed, regardless of the memory tracking mechanism used, when a trim range is reached, GFXR will dump _all_ allocated buffers and images to the file. This is necessary because outside of trimmed region we do not capture commands, so buffer content can not be regenerated at replay by the gpu, so all content needs to be dumped into the file.
In the capture file these writes are visible as `InitBufferCommand` and `InitImageCommand`. These commands are trim capture specific and are absent from un-trimmed captures. In trimmed captures `FillMemoryCommand`s are only present inside the capture range, when user writes are tracked again

**Comment by davidlunarg** [comment]
Hmm... Recently, during a brainstorm, I proposed a simple way to do trimming, knowing full well it's not the optimal way to do trimming - remove all vkQueuePresentKHR calls except for the frame(s) of interest. For most Vulkan programs, this would result in a trimmed file about the same size as the original, with about the same number of Vulkan API calls. However, the trim file produced using this simple trimming method would not reduce the number of Vulkan API calls or execution time relative to the untrimmed file.

But here we have a case where my simple trimming proposal would result in a smaller trimmed file than our current implementation. So if the need should arise in the future to optimize trimming on file size rather than number of API calls or execution time, we know how to do it. :-)

Closing this issue.
